File: internal/rtk/domain.go
package rtk

import (
	"fmt"
	"strings"
)

var (
	materialSeverities = map[string]bool{"high": true, "medium": true}
	allSeverities      = map[string]bool{"high": true, "medium": true, "low": true}
	findingBasis       = map[string]bool{"supported": true, "gap": true}
	decisions          = map[string]bool{"accept": true, "reject": true}
)

func invariant(condition bool, message string) error {
	if !condition {
		return fmt.Errorf("%s", message)
	}
	return nil
}

func validateString(value string, path string) error {
	if strings.TrimSpace(value) == "" {
		return fmt.Errorf("%s must be a non-empty string", path)
	}
	return nil
}

func ValidateProposal(proposal *Proposal, path string) error {
	if path == "" {
		path = "proposal"
	}
	if proposal == nil {
		return fmt.Errorf("%s must be an object", path)
	}
	if err := validateString(proposal.Summary, path+".summary"); err != nil {
		return err
	}
	return nil
}

func ValidateEvidence(evidence *Evidence, path string) error {
	if path == "" {
		path = "evidence"
	}
	if evidence == nil {
		return fmt.Errorf("%s must be an object", path)
	}
	if err := validateString(evidence.ID, path+".id"); err != nil {
		return err
	}
	if err := validateString(evidence.Source, path+".source"); err != nil {
		return err
	}
	if err := validateString(evidence.Kind, path+".kind"); err != nil {
		return err
	}
	if err := validateString(evidence.Phase, path+".phase"); err != nil {
		return err
	}
	if err := validateString(evidence.Statement, path+".statement"); err != nil {
		return err
	}
	if err := validateString(evidence.Excerpt, path+".excerpt"); err != nil {
		return err
	}
	if err := validateString(evidence.CollectedBy, path+".collected_by"); err != nil {
		return err
	}
	if err := invariant(evidence.Round >= 0, path+".round must be a non-negative integer"); err != nil {
		return err
	}
	return validateString(evidence.CreatedAt, path+".created_at")
}

func ValidateFinding(finding *Finding, evidenceIndex map[string]Evidence, path string) error {
	if path == "" {
		path = "finding"
	}
	if finding == nil {
		return fmt.Errorf("%s must be an object", path)
	}
	if err := validateString(finding.ID, path+".id"); err != nil {
		return err
	}
	if err := validateString(finding.Critic, path+".critic"); err != nil {
		return err
	}
	if err := validateString(finding.Summary, path+".summary"); err != nil {
		return err
	}
	if err := validateString(finding.Rationale, path+".rationale"); err != nil {
		return err
	}
	if err := validateString(finding.SuggestedChange, path+".suggested_change"); err != nil {
		return err
	}
	if !allSeverities[finding.Severity] {
		return fmt.Errorf("%s.severity must be one of high|medium|low", path)
	}
	if !findingBasis[finding.Basis] {
		return fmt.Errorf("%s.basis must be one of supported|gap", path)
	}
	if finding.Basis == "supported" && len(finding.EvidenceIDs) == 0 {
		return fmt.Errorf("%s.evidence_ids must be non-empty for supported findings", path)
	}
	if finding.Basis == "gap" && len(finding.EvidenceIDs) != 0 {
		return fmt.Errorf("%s.evidence_ids must be empty for gap findings", path)
	}
	for index, evidenceID := range finding.EvidenceIDs {
		if err := validateString(evidenceID, fmt.Sprintf("%s.evidence_ids[%d]", path, index)); err != nil {
			return err
		}
		if _, ok := evidenceIndex[evidenceID]; !ok {
			return fmt.Errorf("%s.evidence_ids[%d] references unknown evidence %s", path, index, evidenceID)
		}
	}
	return nil
}

func ValidateDecision(decision *Decision, findingIndex map[string]Finding, evidenceIndex map[string]Evidence, path string) error {
	if path == "" {
		path = "decision"
	}
	if decision == nil {
		return fmt.Errorf("%s must be an object", path)
	}
	if err := validateString(decision.FindingID, path+".finding_id"); err != nil {
		return err
	}
	if !decisions[decision.Disposition] {
		return fmt.Errorf("%s.disposition must be one of accept|reject", path)
	}
	if err := validateString(decision.Rationale, path+".rationale"); err != nil {
		return err
	}
	finding, ok := findingIndex[decision.FindingID]
	if !ok {
		return fmt.Errorf("%s.finding_id references unknown finding %s", path, decision.FindingID)
	}
	if finding.Basis == "supported" && len(decision.EvidenceIDs) == 0 {
		return fmt.Errorf("%s.evidence_ids must be non-empty for supported findings", path)
	}
	for index, evidenceID := range decision.EvidenceIDs {
		if err := validateString(evidenceID, fmt.Sprintf("%s.evidence_ids[%d]", path, index)); err != nil {
			return err
		}
		if _, ok := evidenceIndex[evidenceID]; !ok {
			return fmt.Errorf("%s.evidence_ids[%d] references unknown evidence %s", path, index, evidenceID)
		}
	}
	return nil
}

func FindingCounts(findings []Finding) Counts {
	counts := Counts{}
	for _, finding := range findings {
		counts.Total++
		switch finding.Severity {
		case "high":
			counts.High++
		case "medium":
			counts.Medium++
		case "low":
			counts.Low++
		}
		if materialSeverities[finding.Severity] {
			counts.Material++
		}
		if finding.Basis == "gap" {
			counts.Gaps++
		}
	}
	return counts
}

func IsMaterialFinding(finding Finding) bool {
	return materialSeverities[finding.Severity]
}

File: src/domain.js
const MATERIAL_SEVERITIES = new Set(["high", "medium"]);
const ALL_SEVERITIES = new Set(["high", "medium", "low"]);
const FINDING_BASIS = new Set(["supported", "gap"]);
const DECISIONS = new Set(["accept", "reject"]);

function invariant(condition, message) {
  if (!condition) {
    throw new Error(message);
  }
}

function isPlainObject(value) {
  return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

function validateString(value, path) {
  invariant(typeof value === "string" && value.trim().length > 0, `${path} must be a non-empty string`);
}

function validateArray(value, path) {
  invariant(Array.isArray(value), `${path} must be an array`);
}

function validateProposal(proposal, path = "proposal") {
  invariant(isPlainObject(proposal), `${path} must be an object`);
  validateString(proposal.summary, `${path}.summary`);
  invariant(
    proposal.claims === undefined || Array.isArray(proposal.claims),
    `${path}.claims must be an array when present`,
  );
  invariant(
    proposal.acceptance === undefined || Array.isArray(proposal.acceptance),
    `${path}.acceptance must be an array when present`,
  );
  return proposal;
}

function validateEvidence(evidence, path = "evidence") {
  invariant(isPlainObject(evidence), `${path} must be an object`);
  validateString(evidence.id, `${path}.id`);
  validateString(evidence.source, `${path}.source`);
  validateString(evidence.kind, `${path}.kind`);
  validateString(evidence.phase, `${path}.phase`);
  validateString(evidence.statement, `${path}.statement`);
  validateString(evidence.excerpt, `${path}.excerpt`);
  validateString(evidence.collected_by, `${path}.collected_by`);
  invariant(Number.isInteger(evidence.round) && evidence.round >= 0, `${path}.round must be a non-negative integer`);
  validateString(evidence.created_at, `${path}.created_at`);
  return evidence;
}

function validateFinding(finding, evidenceIndex, path = "finding") {
  invariant(isPlainObject(finding), `${path} must be an object`);
  validateString(finding.id, `${path}.id`);
  validateString(finding.critic, `${path}.critic`);
  validateString(finding.summary, `${path}.summary`);
  validateString(finding.rationale, `${path}.rationale`);
  validateString(finding.suggested_change, `${path}.suggested_change`);
  invariant(ALL_SEVERITIES.has(finding.severity), `${path}.severity must be one of high|medium|low`);
  invariant(FINDING_BASIS.has(finding.basis), `${path}.basis must be one of supported|gap`);
  validateArray(finding.evidence_ids, `${path}.evidence_ids`);

  if (finding.basis === "supported") {
    invariant(finding.evidence_ids.length > 0, `${path}.evidence_ids must be non-empty for supported findings`);
  } else {
    invariant(finding.evidence_ids.length === 0, `${path}.evidence_ids must be empty for gap findings`);
  }

  for (const [index, evidenceId] of finding.evidence_ids.entries()) {
    validateString(evidenceId, `${path}.evidence_ids[${index}]`);
    invariant(evidenceIndex.has(evidenceId), `${path}.evidence_ids[${index}] references unknown evidence ${evidenceId}`);
  }

  return finding;
}

function validateDecision(decision, findingIndex, evidenceIndex, path = "decision") {
  invariant(isPlainObject(decision), `${path} must be an object`);
  validateString(decision.finding_id, `${path}.finding_id`);
  invariant(DECISIONS.has(decision.disposition), `${path}.disposition must be one of accept|reject`);
  validateString(decision.rationale, `${path}.rationale`);
  validateArray(decision.evidence_ids, `${path}.evidence_ids`);

  const finding = findingIndex.get(decision.finding_id);
  invariant(finding, `${path}.finding_id references unknown finding ${decision.finding_id}`);

  if (finding.basis === "supported") {
    invariant(decision.evidence_ids.length > 0, `${path}.evidence_ids must be non-empty for supported findings`);
  }

  for (const [index, evidenceId] of decision.evidence_ids.entries()) {
    validateString(evidenceId, `${path}.evidence_ids[${index}]`);
    invariant(evidenceIndex.has(evidenceId), `${path}.evidence_ids[${index}] references unknown evidence ${evidenceId}`);
  }

  return decision;
}

function findingCounts(findings) {
  return findings.reduce(
    (acc, finding) => {
      acc.total += 1;
      acc[finding.severity] += 1;
      if (MATERIAL_SEVERITIES.has(finding.severity)) {
        acc.material += 1;
      }
      if (finding.basis === "gap") {
        acc.gaps += 1;
      }
      return acc;
    },
    { total: 0, high: 0, medium: 0, low: 0, material: 0, gaps: 0 },
  );
}

function isMaterialFinding(finding) {
  return MATERIAL_SEVERITIES.has(finding.severity);
}

module.exports = {
  findingCounts,
  invariant,
  isMaterialFinding,
  validateDecision,
  validateEvidence,
  validateFinding,
  validateProposal,
};
